Add more tips on writing native applications (#26676)
quarkusio/quarkus: docs/src/main/asciidoc/writing-native-applications-tips.adoc

Solving this issue requires adding the `-H:DynamicProxyConfigurationResources=<comma-separated-config-resources>` option and to provide a dynamic proxy configuration file.
You can find all the information about the format of this file in https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/DynamicProxy.md#manual-configuration[the GraalVM documentation].

[[modularity-benefits]]
=== Modularity Benefits

As already explained in sections above,
during native image build time GraalVM analyses the application's call tree and generates a universe of all the code it needs.
Having a modular codebase is key to avoiding problems with unused or optional parts of your application,

> while at the same time reducing both native image build times and native executable size.

Reviewer: ```suggestion
while at the same time reducing both native executable build times and size.
```

In this section you will learn about the details behind the benefits of modularity for native applications.

When code is not modular enough, generated native executables can end up with more code than what the user needs.
If a feature is not used and the code gets compiled into the native executable,

> this is both a waste of native compilation time and native executable disk space.

Reviewer 2: Isn't it also a waste of memory at the Quarkus level, given we initialize most classes at build time?

Even more problems arise when third party libraries or specialized API subsystems are used which cause native compilation or runtime errors,
and their use is not modularised enough.

> A recent problem can be found in the jaxb library,

Reviewer 2: ```suggestion
A recent problem can be found in the JAXB library,
```
Please do a search and replace for this one.

which is capable of deserializing XML files containing images using Java’s AWT APIs.
The vast majority of Quarkus XML users don’t need to deserialize images,
so there shouldn’t be a need for users applications to include Java AWT code,

> unless they would specifically configure Quarkus to add the jaxb AWT code to the native executable.

Reviewer 3: `unless they specifically configure... `

However, because jaxb code that uses AWT is in the same jar as the rest of the XML parsing code,
achieving this separation was rather complex and required the use of Java bytecode substitutions to get around it.

> These substitutions are hard to maintain and can easily break, hence they are really the last line of defence.

Author: Sounds good. I think there's a small apostrophe missing (`ones` -> `one's`)

A modular codebase is the best way to avoid these kind of issues.
Taking the jaxb/AWT problem above,
if the jaxb code that dealt with images was in a separate module or jar (e.g. `jaxb-images`),
then Quarkus could choose not to include that module unless the user specifically requested the need to serialize/deserialize XML files containing images at build time.

An added benefit of modular applications can make the universe of code that gets into the native executable smaller.
The smaller the universe, the faster it native image builds will be and the smaller the native executable produced.

The key takeaway point here is the following:
Keeping optional features, particularly those that depend on third party libraries or API subsystems with a big footprint,

> in separate modules is the best solution.

Reviewer: This might be worth putting in an asciidoc block.

Author: I'm wrapping this in a tip block. Is that what you had in mind?

Reviewer: Yes. Thank you

> As a final note,
> how do I know if my application could suffer from similar problems?

Reviewer: ```suggestion
How do I know if my application suffers from similar problems?
```

Aside from a deep study of the application,
finding usages of
https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html[Maven optional dependencies]
is a clear indicator that your application might suffer from similar problems.
These type of dependencies should be avoided,
and instead code that interacts with optional dependencies should be moved into separate modules.

[[enforcing-singletons]]
=== Enforcing Singletons

As already explained in the <<delay-class-init, delay class initialization>> section,
Quarkus marks all code to be initialized at build time unless otherwise specified.
This means that, unless marked otherwise,
static variables will be assigned at build time,
and static blocks will be executed at build time too.

> This can cause values in Java programs that would vary from one run to another,

Reviewer: ```suggestion
This can cause values in Java programs that would normally vary from one run to another,
```

to always return a constant value.
E.g. a static field that is assigned the value of `System.currentTimeMillis()`

> will always return the same value when executed as Quarkus native.

Reviewer 2: ```suggestion
will always return the same value when executed as a Quarkus native executable.
```

Singletons that rely on either static variable initialization will suffer similar problems.
For example, imagine you have a singleton based around static initialization along with a REST endpoint to query it:

[source,java]
----
@Path("/singletons")
public class Singletons {

    @GET
    @Path("/static")
    public long withStatic() {
        return StaticSingleton.startTime();
    }
}

class StaticSingleton {
    static final long START_TIME = System.currentTimeMillis();

    static long startTime() {
        return START_TIME;
    }
}
----

When the `singletons/static` endpoint is queried,
it will always return the same value,
even after the application is restarted:

[source,bash]
----
$ curl http://localhost:8080/singletons/static
1656509254532%

$ curl http://localhost:8080/singletons/static
1656509254532%

### Restart the native application ###

$ curl http://localhost:8080/singletons/static
1656509254532%
----

> Singletons that rely on `enum` classes also show the same issue:

Reviewer: ```suggestion
Singletons that rely on `enum` classes are also affected by the same issue:
```

[source,java]
----
@Path("/singletons")
public class Singletons {

    @GET
    @Path("/enum")
    public long withEnum() {
        return EnumSingleton.INSTANCE.startTime();
    }
}

enum EnumSingleton {
    INSTANCE(System.currentTimeMillis());

    private final long startTime;

    private EnumSingleton(long startTime) {
        this.startTime = startTime;
    }

    long startTime() {
        return startTime;
    }
}
----

When the `singletons/enum` endpoint is queried,
it will always return the same value,
even after the application is restarted:

[source,bash]
----
$ curl http://localhost:8080/singletons/enum
1656509254601%

$ curl http://localhost:8080/singletons/enum
1656509254601%

### Restart the native application ###

$ curl http://localhost:8080/singletons/enum
1656509254601%
----

> Both of the examples above can be fixed to behave like a singleton in the JVM mode by marking the singleton classes to be runtime initialized.

Reviewer: ```suggestion
Both of the examples above can be fixed to behave like they would in JVM mode by marking the singleton classes to be runtime initialized.
```

Reviewer 2: A link to how it can be done could be useful?

Author: Hmmm sure I can do. Just note how to do this is already explained earlier in the very same document.

Author: Actually, I'm removing this line because just after the cdi singleton example curl examples this recommendation is duplicated and there I include a link (to the wrong place, which I'm fixing).

One more way to fix it is to build singletons using CDI's `@Singleton` annotation:

[source,java]
----
@Path("/singletons")
public class Singletons {

    @Inject
    CdiSingleton cdiSingleton;

    @GET
    @Path("/cdi")
    public long withCdi() {
        return cdiSingleton.startTime();
    }
}

@Singleton
class CdiSingleton {

>     final long startTime = System.currentTimeMillis();

Reviewer 2: I think it's important to add a comment that the field is not static here. Because if it was, you would have the same issue. Maybe with a callout?

    long startTime() {
        return startTime;
    }
}
----

After each restart,
querying `singletons/cdi` will return a different value,
just like it would in JVM mode:

[source,bash]
----
$ curl http://localhost:8080/singletons/cdi
1656510218554%

$ curl http://localhost:8080/singletons/cdi
1656510218554%

### Restart the native application ###

$ curl http://localhost:8080/singletons/cdi
1656510714689%
----

One more simple way to enforce a singleton relying static fields, or enums,
is to <<delay-class-init-in-your-app,delay its class initialization until run time>>.
The nice advantage of CDI-based singletons is that your class initialization is not constrained,
so you can freely decide whether it should be build-time or run-time initialized,

> depending on your use case.

Reviewer 2: Isn't it the first thing you mentioned above before the singleton example? It looks weird to come back to it now with `One more simple way`.

Author: Yeah there was some duplication. I'm removing the above mention and I'm rewording this paragraph.

=== Beware of common Java API overrides

Certain commonly used Java methods are overriden by user classes,

> e.g. `toString`, `equals`, `hashCode`...etc.

Reviewer 2: I think you should make it extra clear that this is relevant only if these methods are doing funky things? Because in most cases, it's not a problem at all?

Author: @gsmet Indeed that's correct. I'll reiterate those points.

@DanHeidinga  It's not only extra code that can be a problem. If implementations do reflection or use proxies, you could get a similar effect where your native image fails because there's some extra configuration missing. I'll add details on this.

Author: @DanHeidinga Oh, about examples. I don't have any right now. I think @Sanne might have since this came from a discussion I had with him a while back. IIRC some `toString()` impls in Hibernate (not sure if in the code itself, or in user provided entities) did something funky that caused the native image build to break.

From a GraalVM points-to analysis perspective,
what happens in these method overrides matters,
even if the application does not explicitly call them.
This is because these methods are used throughout the JDK,

> and all it takes it's for one of those calls to be done on an unconstrained type,

Reviewer 2: ```suggestion
and all it takes is for one of those calls to be done on an unconstrained type,
```

e.g. `java.lang.Object`,
for the analysis to have to pull all implementations of that particular method.

> So, what can you if the implementation one of these methods is giving trouble during native executable build time?

Reviewer: ```suggestion
So, what can you do if the implementation of one of these methods is causing issues during native executable build time?
```

The best thing would be to come up with parallel methods that don't override super class methods.
E.g. instead of defining `String toString() { ... }` method,

> create a `String show() { ... }` method.

Reviewer 2: Well, that could work for `toString()` but will probably be a pain but is definitely not useful for hashCode and equals.
TBH, I'm not sure if this paragraph brings a lot of value given in most cases these methods will be very simple and won't bring any new classes.
If we're worried it could be a problem, I think we need to be extremely clear at the beginning of the paragraph and describe exactly when it's a real problem.

Author: +1 to both. I'll remove this refactoring suggestion.

If you can't modify the source code to apply this suggestion,
you might be able to use GraalVM substitutions,
but these are quite brittle and can break easily.
So if you end up having to use these,

> you should try use them only temporarily,

Reviewer: ```suggestion
you should try to use them only temporarily,
```

> until you have the chance to rework/rewrite the code to avoid the original issue.

Reviewer: ```suggestion
as a work-around until you have the chance to rework/rewrite the code to avoid the original issue.
```

The best way to find out what parts of the source code get included in the analysis

> is by looking at the <<native-reference.adoc#native-reports,native build time reports>>.

Reviewer: It's not clear how this sentence relates to the above.

When code is not modular enough, generated native executables can end up with more code than what the user needs.
If a feature is not used and the code gets compiled into the native executable,
this is a waste of native compilation time and memory usage, as well as native executable disk space and starting heap size.
Even more problems arise when third party libraries or specialized API subsystems are used which cause native compilation or runtime errors,
and their use is not modularised enough.
A recent problem can be found in the JAXB library,
which is capable of deserializing XML files containing images using Java’s AWT APIs.
The vast majority of Quarkus XML users don’t need to deserialize images,
so there shouldn’t be a need for users applications to include Java AWT code,
unless they specifically configure Quarkus to add the JAXB AWT code to the native executable.
However, because JAXB code that uses AWT is in the same jar as the rest of the XML parsing code,
achieving this separation was rather complex and required the use of Java bytecode substitutions to get around it.

> These substitutions are hard to maintain and can easily break, hence they should be the one's last resort.

Reviewer: ```suggestion
These substitutions are hard to maintain and can easily break, hence they should be one's last resort.
```

then Quarkus could choose not to include that module unless the user specifically requested the need to serialize/deserialize XML files containing images at build time.

Another benefit of modular applications is that they can reduce the code-set that will need to get into the native executable.
The smaller the code-set, the faster the native executable builds will be and the smaller the native executable produced.

[TIP]
====
The key takeaway point here is the following:
Keeping optional features, particularly those that depend on third party libraries or API subsystems with a big footprint,
in separate modules is the best solution.
====

How do I know if my application suffers from similar problems?
Aside from a deep study of the application,
finding usages of
https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html[Maven optional dependencies]
is a clear indicator that your application might suffer from similar problems.
These type of dependencies should be avoided,
and instead code that interacts with optional dependencies should be moved into separate modules.

[[enforcing-singletons]]
=== Enforcing Singletons

> As already explained in the <<delay-class-init, delay class initialization>> section,
> Quarkus marks all code to be initialized at build time unless otherwise specified.
> This means that, unless marked otherwise,

Reviewer 4: ```suggestion
As already explained in the <<delay-class-init, delay class initialization>> section,
Quarkus marks all code to be initialized at build time by default.
This means that, unless marked otherwise,
```

Minor nitpick on phrasing to avoid repeated use of "otherwise"

even after the application is restarted:

[source,bash]
----
$ curl http://localhost:8080/singletons/static
1656509254532%

$ curl http://localhost:8080/singletons/static
1656509254532%

### Restart the native application ###

$ curl http://localhost:8080/singletons/static
1656509254532%
----

Singletons that rely on `enum` classes are also affected by the same issue:

[source,java]
----
@Path("/singletons")
public class Singletons {

    @GET
    @Path("/enum")
    public long withEnum() {
        return EnumSingleton.INSTANCE.startTime();
    }
}

enum EnumSingleton {
    INSTANCE(System.currentTimeMillis());

    private final long startTime;

    private EnumSingleton(long startTime) {
        this.startTime = startTime;
    }

    long startTime() {
        return startTime;
    }
}
----

When the `singletons/enum` endpoint is queried,
it will always return the same value,
even after the application is restarted:

[source,bash]
----
$ curl http://localhost:8080/singletons/enum
1656509254601%

$ curl http://localhost:8080/singletons/enum
1656509254601%

### Restart the native application ###

$ curl http://localhost:8080/singletons/enum
1656509254601%
----

> One more way to fix it is to build singletons using CDI's `@Singleton` annotation:

Reviewer 4: ```suggestion
One way to fix it is to build singletons using CDI's `@Singleton` annotation:
```

Removed "more" as this is the first proposed fix.
